src/dinitctl.cc

```diff
@@ -1903,38 +1903,43 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, service_dir_
     }
 
     if (socknum >= 0) {
+        if (verbose) {
+            cout << "Service '" << to << "' has been " << (enable ? "enabled" : "disabled") << "." << endl;
+        }
+
         // Check status of the service now
         auto m = membuf()
                 .append((char)cp_cmd::SERVICESTATUS)
                 .append(to_handle);
         write_all_x(socknum, m);
 
-        wait_for_reply(rbuffer, socknum);
-        if (rbuffer[0] != (char)cp_rply::SERVICESTATUS) {
-            cerr << "dinitctl: protocol error." << endl;
-            return 1;
+        try {
+            wait_service_state(socknum, rbuffer, to_handle, to, !enable, verbose);
         }
-        rbuffer.consume(1);
-
-        int statussize = 6 + std::max(sizeof(pid_t), sizeof(int));;
-        fill_buffer_to(rbuffer, socknum, statussize + 1 /* reserved */);
-        rbuffer.consume(1);
-        service_state_t current = static_cast<service_state_t>(rbuffer[0]);
-        service_state_t target = static_cast<service_state_t>(rbuffer[1]);
-        rbuffer.consume(statussize);
+        catch (const dinit_protocol_error &) {
+            // We probably got a SERVICESTATUS before SERVICEEVENT, Process it.
```

Comment:
Do not use exceptions for ordinary control flow. https://isocpp.org/wiki/faq/exceptions#why-not-exceptions

`wait_service_state` throws a protocol error if it sees an unexpected reply packet because it is not supposed to be used when expecting a reply packet, and so if it does see a reply packet then that is an error condition. Relying on the buffer still being valid and still containing the reply packet that was received is encoding an unstated assumption about the behaviour of `wait_service_state` which should not be relied on.

Reply:
Right. So the correct way to do that is this?
```
        fill_buffer_to(rbuffer, socknum, 1);
        if (rbuffer[0] >= 100) {
            // We got a SERVICEEVENT before the SERVICESTATUS reply.
            wait_service_state(socknum, rbuffer, to_handle, to, !enable, verbose);
        }
        else if (rbuffer[0] == (char)cp_rply::SERVICESTATUS) {
            ...
        }
        else {
            // protocol error
        }
```

Reply:
Packet ids >= 100 are reserved for information packets that can be sent at any time (out-of-band). The only type of information packet is currently `SERVICEEVENT` but that might change in the future, so it is not a good idea to assume that receiving a packet with id >= 100 means that a `SERVICEEVENT` will necessarily be sent, either.

Reply:
It's fixed now. I thought `ids >= 100` is OK but I just missing the second statement in `wait_service_state()` function:
https://github.com/davmac314/dinit/blob/3867cf1766134980d2c3cd6f441276217af498e9/src/dinitctl.cc#L787-L791

Reply:
Information packets that are not `SERVICEEVENT` must be correctly handled, as I mentioned:

> Packet ids >= 100 are reserved for information packets that can be sent at any time (out-of-band).